gnd/src/main/java/com/google/android/gnd/ui/editrecord/EditRecordViewModel.java

```diff
@@ -78,6 +86,47 @@
     this.showUnsavedChangesDialogEvents = new SingleLiveEvent<>();
     this.showErrorDialogEvents = new SingleLiveEvent<>();
     this.authManager = authenticationManager;
+    this.editRecordRequests = PublishSubject.create();
+    this.recordSaveRequests = PublishSubject.create();
+
+    disposeOnClear(
+            recordSaveRequests.zipWith(authManager.getUser(), this::saveRecord)
+            .switchMap(Result::wrapObservable)
+            .subscribe(Result.unwrap(record::setValue, this::onSaveRecordError)));
+
+    disposeOnClear(
+        editRecordRequests
+            .switchMapSingle(Result.mapSingle(this::createOrUpdateRecord))
+            .subscribe(Result.unwrap(this::onRecordSnapshot, this::onEditRecordError)));
+  }
+
+  private Single<Resource<Record>> createOrUpdateRecord(EditRecordRequest request) {
+    return request.getIsNew() ? createRecord(request) : updateRecord(request);
+  }
+
+  private Single<Resource<Record>> createRecord(EditRecordRequest request) {
+    return this.dataRepository
+        .createRecord(request.getProjectId(), request.getFeatureId(), request.getFormId())
+        .map(Resource::loaded)
+        .doFinally(this::onNewRecordLoaded);
+  }
+
+  private Single<Resource<Record>> updateRecord(EditRecordRequest request) {
+    return this.dataRepository
+        .getRecordSnapshot(request.getProjectId(), request.getFeatureId(), request.getRecordId())
+        .doOnSuccess(r -> r.data().ifPresent(this::update));
```

Comment:
I assume this is out of scope of this PR, but eventually would be nice to stop relying on side-effects, e.g. can we call these in `onRecordSnapshot()` instead. We can get back to this later.

Reply:
Sounds good, #78